storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -24,6 +26,175 @@ import (
+and then reschedule the request to monitor the evolution of the job.
+
+If we do have a job ID, then something is in progress and we monitor it until
+it's over.
+If it has ended with an error, we move the volume into a failed state.
+Otherwise we proceed with the PersistentVolume creation (which may require an
+extra Salt call, synchronous this time, to get the disk size), taking care of
+putting a finalizers on both ourself and the PersistentVolume (so that our
+lifetimes are tied) and setting ourself as the onwer of the PersistentVolume.
+
+Once we have successfuly created the PersistentVolume, we can move into the
+`Available` state and reschedule the request (the next iteration will check the
+health of the PersistentVolume we just created).
+
+                    +------------------+             +----------+
+              +---->|SpawnPrepareVolume|------------>|SetPending|
+              |     +------------------+             +----------+
+              | NO                                        |
+              |                                           v
+ -----     +----+ DONE  +--------+   +------------+   +-------+    ------
+(START)--->|Job?|------>|CreatePV|-->|SetAvailable|-->|Requeue|-->( STOP )
+ -----     +----+       +--------+   +------------+   +-------+    ------
+             | YES                                         ^
+             v                                             |
+        +-----------+ Job Failed    +---------+            |
+        |           |-------------->|SetFailed|----------->+
+        |           |               +---------+            |
+        |           |                                      |
+        |           | Unknown Job   +--------+             |
+        |PollSaltJob|-------------->|UnsetJob|------------>+
+        |           |               +--------+             |
+        |           |                                      |
+        |           | Job Succeed   +--------+             |
+        |           |-------------->|Job=DONE|------------>+
+        +-----------+               +--------+             |
+             | Job in progress                             |
+             |                                             |
+             +---------------------------------------------+
+
+
+
+===================
+= Finalize Volume =
+===================
+
+`Pending` volumes cannot be deleted (because we don't know where we are in the
+creation process), so we reschedule the request until the volume become either
+`Failed` or `Available`.
+
+For volumes with no backing PersistentVolume (the creation failed, we already
+deleted it, …), we simply remove the `volume-protection` finalizer, since they
+are `Terminating` they will be deleted by Kubernetes once the finalizer is gone.
+
+If we do have a backing PersistentVolume, we need to check if we can delete it.
+If not (because it is bound for instance), then we reschedule the request to
```

Comment:
Why can't we simply delete it in this case? If the `Volume` is deleted, we should `DELETE` the `PersistentVolume`. The existing finalizers will ensure it doesn't actually gets deleted.

Reply:
Finalization flow was reworked to address this aspect